core/services/gateway/handlers/functions/allowlist.go

```diff
@@ -0,0 +1,130 @@
+package functions
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+
+	evmclient "github.com/smartcontractkit/chainlink/v2/core/chains/evm/client"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/job"
+)
+
+// OnchainAllowlist maintains an allowlist of addresses that if fetched from the blockchain (EVM-only).
+// Updating the allowlist can be achieved by either manually calling the UpdateFromChain method
+// or starting the service (Start()/Close()), which will periodically update the allowlist.
+//
+//go:generate mockery --quiet --name OnchainAllowlist --output ./mocks/ --case=underscore
+type OnchainAllowlist interface {
+	job.ServiceCtx
+
+	Allow(common.Address) bool
+	UpdateFromChain() error
+}
+
+type onchainAllowlist struct {
+	client          evmclient.Client
+	allowlist       map[common.Address]struct{}
+	mu              sync.RWMutex
+	contractAddress common.Address
+	abiReturnType   abi.Arguments
+	methodSignature []byte
+	updateFrequency time.Duration
+	shutdownCh      chan struct{}
+	lggr            logger.Logger
+}
+
+const (
+	AllowlistArgumentType = "address[]"
+	AllowlistMethodHeader = "getAuthorizedSenders()"
+)
+
+func NewOnchainAllowlist(client evmclient.Client, contractAddress common.Address, updateFrequency time.Duration, lggr logger.Logger) (OnchainAllowlist, error) {
+	if client == nil {
+		return nil, errors.New("client is nil")
+	}
+	abiType, err := abi.NewType(AllowlistArgumentType, "", []abi.ArgumentMarshaling{})
+	if err != nil {
+		return nil, fmt.Errorf("unexpected error during abi.NewType: %s", err)
+	}
+	abiArgs := abi.Arguments([]abi.Argument{{Name: "addresses", Type: abiType}})
+	methodSignature := crypto.Keccak256([]byte(AllowlistMethodHeader))[:4]
+	return &onchainAllowlist{
+		client:          client,
+		allowlist:       make(map[common.Address]struct{}),
+		contractAddress: contractAddress,
+		abiReturnType:   abiArgs,
+		methodSignature: methodSignature,
+		updateFrequency: updateFrequency,
+		shutdownCh:      make(chan struct{}),
+		lggr:            lggr.Named("OnchainAllowlist"),
+	}, nil
+}
+
+func (a *onchainAllowlist) Allow(address common.Address) bool {
+	a.mu.RLock()
+	defer a.mu.RUnlock()
+	_, ok := a.allowlist[address]
+	return ok
+}
+
+func (a *onchainAllowlist) UpdateFromChain() error {
+	rawResponse, err := a.client.CallContract(context.Background(), ethereum.CallMsg{
```

Comment:
Does the core node not have any simplified utility functions which allows you to specify the function ABI and return type ABI as strings to perform the contract call & decode the response?  This seems like it may be useful elsewhere.